backend/repository/thrift_compile.go

```diff
@@ -407,3 +434,278 @@ func newEnumSpec(enum *compile.EnumSpec) *EnumSpec {
 	}
 	return e
 }
+
+// Code converts the module as normal thrift code.
+func (m *Module) Code() string {
+	codeBlocks := make([]*CodeBlock, 0, 100)
+	for _, ns := range m.Namespace {
+		codeBlocks = append(codeBlocks, ns.CodeBlock())
+	}
+	for _, im := range m.Includes {
+		codeBlocks = append(codeBlocks, im.CodeBlock(m.ThriftPath))
+	}
+	for _, c := range m.Constants {
+		codeBlocks = append(codeBlocks, c.CodeBlock(m.ThriftPath))
+	}
+	for _, t := range m.Types {
+		codeBlocks = append(codeBlocks, t.CodeBlock(m.ThriftPath))
+	}
+	for _, service := range m.Services {
+		codeBlocks = append(codeBlocks, service.CodeBlock(m.ThriftPath))
+	}
+	sort.Sort(codeBlockSlice(codeBlocks))
+	result := bytes.NewBuffer(nil)
+	for i := range codeBlocks {
+		result.WriteString(codeBlocks[i].Code)
+		result.WriteString("\n\n")
+	}
+	return result.String()
+}
+
+// CodeBlock defines a code block of a thrift file.
+type CodeBlock struct {
+	Code  string
+	Order int
+}
+
+// CodeBlock converts a namespace to a CodeBlock.
+func (ns *Namespace) CodeBlock() *CodeBlock {
+	return &CodeBlock{
+		Code:  fmt.Sprintf("namespace %s %s", ns.Scope, ns.Name),
+		Order: ns.Line,
+	}
+}
+
+// CodeBlock converts a included module to a CodeBlock.
+func (im *IncludedModule) CodeBlock(curFilePath string) *CodeBlock {
+	relPath, err := filepath.Rel(curFilePath, im.Module.ThriftPath)
+	if err != nil {
+		relPath = curFilePath
+	}
+	relPath = filepath.Clean(strings.TrimPrefix(relPath, "../"))
+	return &CodeBlock{
+		Code:  fmt.Sprintf("include \"%s\"", relPath),
+		Order: im.Line,
+	}
+}
+
+// CodeBlock converts a constant to a CodeBlock.
+func (c *Constant) CodeBlock(curFilePath string) *CodeBlock {
+	return &CodeBlock{
+		Code:  fmt.Sprintf("const %s %s = %s", c.Type.FullTypeName(curFilePath), c.Name, c.Value),
+		Order: c.Line,
+	}
+}
+
+// CodeBlock converts a typeSpec to a CodeBlock.
+func (ts *TypeSpec) CodeBlock(curFilePath string) *CodeBlock {
+	// Typedef definition
+	if ts.TypeDefTarget != nil {
+		return &CodeBlock{
+			Code: fmt.Sprintf("typedef %s %s", ts.Name,
+				ts.TypeDefTarget.FullTypeName(curFilePath)),
+			Order: ts.Line,
+		}
+	}
+	// Enum definition
+	if ts.EnumType != nil {
+		return enumTypeCodeBlock(ts.EnumType, ts.Line)
+	}
+	// Struct definition
+	if ts.StructType != nil {
+		return structTypeCodeBlock(ts.StructType, ts.Name, ts.Line, curFilePath)
+	}
+	panic(fmt.Sprintf(
+		"No typedef, enum, nor struct definition found: typespec %+v", ts))
+}
+
+func enumTypeCodeBlock(e *EnumSpec, line int) *CodeBlock {
+	result := bytes.NewBuffer(nil)
+	result.WriteString(fmt.Sprintf("enum %s {\n", e.Name))
+	items := e.Items
+	for i := 0; i < len(items)-1; i++ {
+		result.WriteString(fmt.Sprintf("\t%s,\n", items[i].Name))
```

Comment:
It looks like the optional value part of an enum item isn't being serialized.

Reply:
The enum definition doesn't have `optional` part.https://git1-us-west.apache.org/repos/asf?p=thrift.git;a=blob;f=tutorial/tutorial.thrift;hb=HEAD#l91